client/views/view_units.cpp

```diff
@@ -434,7 +490,200 @@ void units_view::upgrade_units()
+{
+  struct unit *best_candidate = NULL;
+  int best_dist = FC_INFINITY, dist;
+
+  players_iterate(pplayer)
+  {
+    if (client_has_player() && pplayer != client_player()) {
+      continue;
+    }
+
+    unit_list_iterate(pplayer->units, punit)
+    {
+      if (utype == unit_type_get(punit)
+          && FOCUS_AVAIL == punit->client.focus_status
+          && 0 < punit->moves_left && !punit->done_moving
+          && punit->ssa_controller == SSA_NONE) {
+        dist = sq_map_distance(unit_tile(punit), ptile);
+        if (dist < best_dist) {
+          best_candidate = punit;
+          best_dist = dist;
+        }
+      }
+    }
+    unit_list_iterate_end;
+  }
+  players_iterate_end;
+
+  return best_candidate;
+}
+
+/**
+ * Returns an array of units data.
+ */
+std::vector<unit_view_entry>
+get_units_view_data(struct unit_view_entry *entries, int *num_entries_used)
+{
+  *num_entries_used = 0;
+
+  if (!client_has_player()) {
+    return {};
+  }
+
+  unit_type_iterate(unittype)
+  {
+    int count = 0;           // Count of active unit type
+    int in_progress = 0;     // Count of being produced
+    int gold_cost = 0;       // Gold upkeep
+    int food_cost = 0;       // Food upkeep
+    int shield_cost = 0;     // Shield upkeep
+    bool upgradable = false; // Unit type is upgradable
+
+    unit_list_iterate(client_player()->units, punit)
+    {
+      if (unit_type_get(punit) == unittype) {
+        count++;
+        upgradable =
+            client_has_player()
+            && nullptr != can_upgrade_unittype(client_player(), unittype);
+
+        // Only units with a home city have upkeep
+        if (punit->homecity != 0) {
+          gold_cost += punit->upkeep[O_GOLD];
+          food_cost += punit->upkeep[O_FOOD];
+          shield_cost += punit->upkeep[O_SHIELD];
+        }
+      }
+    }
+    unit_list_iterate_end;
+
+    city_list_iterate(client_player()->cities, pcity)
+    {
+      if (pcity->production.value.utype == unittype
+          && pcity->production.kind == VUT_UTYPE) {
+        in_progress++;
+      }
+    }
+    city_list_iterate_end;
+
+    // Skip unused unit types
+    if (count == 0 && in_progress == 0) {
+      continue;
+    }
+
+    entries[*num_entries_used].type = unittype;
+    entries[*num_entries_used].count = count;
+    entries[*num_entries_used].in_prod = in_progress;
+    entries[*num_entries_used].upg = upgradable;
+    entries[*num_entries_used].gold_cost = gold_cost;
+    entries[*num_entries_used].food_cost = food_cost;
+    entries[*num_entries_used].shield_cost = shield_cost;
+    (*num_entries_used)++;
+  }
+  unit_type_iterate_end;
+
+  std::sort(entries, entries + *num_entries_used,
+            [](const auto &lhs, const auto &rhs) {
+              return QString::localeAwareCompare(
+                         utype_name_translation(lhs.type),
+                         utype_name_translation(rhs.type))
+                     < 0;
+            });
+
+  return {};
```

Comment:
Guess you shouldn't use `std::vector` if you're anyway always returning an empty one